candy-machine/program/src/lib.rs

```diff
@@ -90,17 +92,26 @@ pub mod candy_machine {
                 return Err(ErrorCode::GatewayTokenMissing.into());
             }
             let gateway_token_info = &ctx.remaining_accounts[remaining_accounts_counter];
-            let gateway_token = ::solana_gateway::borsh::try_from_slice_incomplete::<
-                ::solana_gateway::state::GatewayToken,
-            >(*gateway_token_info.data.borrow())?;
-            // stores the expire_time before the verification, since the verification
-            // will update the expire_time of the token and we won't be able to
-            // calculate the creation time
-            let expire_time = gateway_token
-                .expire_time
-                .ok_or(ErrorCode::GatewayTokenExpireTimeInvalid)?
-                as i64;
             remaining_accounts_counter += 1;
+
+            // Date passed in CPI to check if the token was created before candy machine go live date.
+            // Expire offset is 10 minutes since gatekeeper expires tokens after that long
+            let min_go_live_date = if let Some(val) = &candy_machine.data.go_live_date {
+                if !candy_machine
+                    .data
+                    .whitelist_mint_settings
+                    .as_ref()
+                    .map(|wl| wl.presale)
+                    .unwrap_or(false)
+                {
+                    Some(val + EXPIRE_OFFSET)
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+
```

Comment:
Replace this with a closure that you will then pass in as an eval function. Something like 
```rust
let eval_function = |token| match (&candy_machine.data, token.expire_time()) {
    (
        CandyMachineData {
            go_live_date: Some(go_live_date),
            whitelist_mint_settings: Some(WhitelistMintSettings { presale, .. }),
            ..
        },
        Some(expire_time),
    ) if presale && expire_time < go_live_date + EXPIRE_OFFSET => {
        Err(ErrorCode::GatewayTokenExpireTimeInvalid.into())
    }
    _ => Ok(()),
};
```